src/bvh/bvh_node.rs

```diff
@@ -0,0 +1,608 @@
+                }
+                if ray.intersects_aabb(child_r_aabb) {
+                    BvhNode::traverse_recursive(nodes, child_r_index, ray, indices);
+                }
+            }
+            BvhNode::Leaf { shape_index, .. } => {
+                indices.push(shape_index);
+            }
+        }
+    }
+}
+
+/// Shapes holds a mutable ptr to the slice of Shapes passed in to build. It is accessed only through a ShapeIndex.
+/// These are a set of unique indices into Shapes that are generated at the start of the build process. Because they
+/// are all unique they guarantee that access into Shapes is safe to do in parallel.
+pub struct Shapes<'a, S> {
+    ptr: *mut S,
+    len: usize,
+    marker: PhantomData<&'a S>,
+}
+
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug)]
+pub(crate) struct ShapeIndex(pub usize);
+
+impl<S> Shapes<'_, S> {
+    pub(crate) fn set_node_index<T: BHValue, const D: usize>(
+        &self,
+        shape_index: ShapeIndex,
+        node_index: usize,
+    ) where
+        S: BHShape<T, D>,
+    {
+        assert!(shape_index.0 < self.len);
+        unsafe {
+            self.ptr
+                .add(shape_index.0)
+                .as_mut()
+                .unwrap()
+                .set_bh_node_index(node_index);
+        }
+    }
+    pub(crate) fn get<T: BHValue, const D: usize>(&self, shape_index: ShapeIndex) -> &S
+    where
+        S: BHShape<T, D>,
+    {
+        assert!(shape_index.0 < self.len);
+        unsafe { self.ptr.add(shape_index.0).as_ref().unwrap() }
+    }
+
+    pub(crate) fn from_slice<T: BHValue, const D: usize>(slice: &mut [S]) -> Shapes<S>
+    where
+        S: BHShape<T, D>,
+    {
+        Shapes {
+            ptr: slice.as_mut_ptr(),
+            len: slice.len(),
+            marker: PhantomData,
+        }
+    }
+}
+
+unsafe impl<S: Send> Send for Shapes<'_, S> {}
+unsafe impl<S> Sync for Shapes<'_, S> {}
+
+/// Holds the arguments for calling build.
+pub struct BvhNodeBuildArgs<'a, S, T: BHValue, const D: usize> {
+    shapes: &'a Shapes<'a, S>,
+    indices: &'a mut [ShapeIndex],
+    nodes: &'a mut [MaybeUninit<BvhNode<T, D>>],
+    parent_index: usize,
+    depth: u32,
+    node_index: usize,
+    aabb_bounds: Aabb<T, D>,
+    centroid_bounds: Aabb<T, D>,
+}
+
+impl<'a, S, T: BHValue, const D: usize> BvhNodeBuildArgs<'a, S, T, D> {
+    /// Creates the args
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        shapes: &'a Shapes<'a, S>,
+        indices: &'a mut [ShapeIndex],
+        nodes: &'a mut [MaybeUninit<BvhNode<T, D>>],
+        parent_index: usize,
+        depth: u32,
+        node_index: usize,
+        aabb_bounds: Aabb<T, D>,
+        centroid_bounds: Aabb<T, D>,
+    ) -> Self {
+        Self {
+            shapes,
+            indices,
+            nodes,
+            parent_index,
+            depth,
+            node_index,
+            aabb_bounds,
+            centroid_bounds,
+        }
+    }
```

Comment:
I'm wondering now that I see the usages of this: Might it be better _not_ to have this `new()`? It hides the names of the parameters and you end up with something like
```rust
BvhNodeBuildArgs::new(&shapes, &mut indices, uninit_slice, 0, 0, 0, aabb, centroid),
```
It might be more obvious in the few cases where you call this to just construct it explicitly with the normal struct construction syntax.

Not saying we need to remove this, just putting it up for discussion.

Reply:
SGTM, I mostly was trying to prevent people from outside the crate from making this struct for safety reason. Just made the fields pub(crate) and I like this more. Thanks for the suggestion.